### ISSUE TITLE
adjust placement of fees in checkout cart

### ISSUE THREAD

**Issue by SDavisMedia** [easydigitaldownloads/easy-digital-downloads#4833]
Right now, all fees display on the checkout cart in the same fashion as products. This presents a few problems I can think of:

1. It gives the impression that fee amount is factored into the `Subtotal`, which displays below the fees.

2. There are different types of fees used for different reasons. Fees that behave like discounts display in a totally different fashion than actual discounts. It's just inconsistent.

3. Nothing about the current display of fees explains what things like taxes or a discount does to the fee itself.

Example, below is a **$100 item** with a **$100 shipping fee** and **6.25% tax** applied. Tax information is set to display on the cart.

![screen shot 2016-07-26 at 11 42 08 am](https://cloud.githubusercontent.com/assets/2359131/17150889/0a8ee084-5326-11e6-9de6-18b303aa2397.png)

The math is correct when we look at the logic. Looking at the cart, though, you have to know what's going on to understand this. For example, how did taxes affect the fee? How taxes affect the product is clear. For fees, not so much.

-----

I'd like to see a restructuring of this information. Here are some initial points:

* If a fee is a non cart item (like an EDD Wallet deposit), it should continue to display where it is but be enhanced to show appropriate information like applied taxes. It should match regular cart item display if it is to be treated as one.

* If a fee is a cart item fee (like a Simple Shipping fee), it should display below the `Subtotal` in a footer row. These fees do not affect the subtotal but instead, make adjustments to the subtotal just like taxes and traditional discounts. 

* (optional, maybe) If a cart item fee is displayed below the subtotal and is affected by taxes or a discount, it should display such information.

-----

The overall goal is the leave nothing to question about the _numbers_ while avoiding information overload. Right now, though, there just isn't enough information. 

Here's a **rough** mockup of what I'm thinking using the same cart pictured above:

![screen shot 2016-07-26 at 11 57 20 am](https://cloud.githubusercontent.com/assets/2359131/17151348/26331ca4-5328-11e6-9c74-740e905dc981.png)

This would require a lot of thought. How to display tax information is the tricky part. It should not be redundant. Likewise, it should not show tax TOTALS prematurely. Showing tax INFO is a different story. 

In fact, to help the cart item price actually agree with the subtotal display, I'd even suggest this display when the display tax in cart setting is on:

![screen shot 2016-07-26 at 12 08 19 pm](https://cloud.githubusercontent.com/assets/2359131/17151766/afdff124-5329-11e6-9c12-af106e23fdf0.png)

(take note that the cart item now shows tax information, not a tax total for the product)

I know this is ugly but I think the display of properly structure cart information should be the first priority.

@easydigitaldownloads/core-devs Thoughts?

**Comment by SDavisMedia** [comment]
Also keep in mind that I used the most basic fee I could apply to make this issue easy to understand. Start talking more complex fees like discounts and you'll see why this is such a big deal.

**Comment by sunnyratilal** [comment]
@SDavisMedia I agree with your reasoning.  The separation is necessary to make sense of the cart. The initial screenshot you used above is a perfect example to have the information restructured.

**Comment by mintplugins** [comment]
I definitely agree. I am constantly thrown off by the subtotal myself. Screenshot #1 I think is my preferred one.

**Comment by cklosowski** [comment]
😍  This is so needed. Would make things much clearer.

**Comment by SDavisMedia** [comment]
Gonna drop a note here until we decide whether or not it's a separate issue:

Above, we've mainly discussed `fee` type fees and how they should adjust the subtotal and be displayed as such. Well, `item` type fees (EDD Wallet deposit) should continue to display like actual products but they do not currently affect the subtotal.

![screen shot 2016-07-26 at 1 06 40 pm](https://cloud.githubusercontent.com/assets/2359131/17153496/d7832ae0-5331-11e6-95da-159a7d47aa28.png)

So this is kind of the same problem but on the other side of the fence.

**Comment by mintplugins** [comment]
Just to include some convo from Slack here, there are 2 different types of fees in EDD:

**- Item Fees**
An item fee is a fee that is an item. It acts like a product and is really more of a product than it is a fee.

**- Fees**
A fee is a cost added to the cart because a specific item is in the cart. This would be something like a Shipping fee because a shippable item is in the cart.

The consensus seems to be that Item Fees should be displayed in the cart like any other product/item. In contrast, fees should be listed below the subtotal.

**Comment by SDavisMedia** [comment]
Great minds think alike. 😃

**Comment by pippinsplugins** [comment]
This is spot on @SDavisMedia.

I feel strongly that this should be made a priority for 2.7. We get _so_ many support tickets related to this.

**Comment by cayci** [comment]
Hi, all.  Do you all have this covered, or would you like for me to work on the implementation?  I've been out of the field for a while, and I badly need the chance to prove my skills in order to be considered for a remote position.  Whatcha think?

**Comment by pippinsplugins** [comment]
We are more than happy to accept patches and suggestions!

**Comment by cayci** [comment]
Awesome!  I will check out prior tickets to see what people are looking for, and then get back to you my suggestions before submitting a PR?  Or should I submit a preliminary PR, and wait for suggestions on improving it?  Sorry for the silly questions... it's been a long time since I've done open source, and I don't want to step on toes  :)

**Comment by cayci** [comment]
"The consensus seems to be that Item Fees should be displayed in the cart like any other product/item. In contrast, fees should be listed below the subtotal."

All in favor?

**Comment by SDavisMedia** [comment]
@cayci That's definitely how I feel about them. Also note that Item Fees need to affect the subtotal since they're basically products. Right now, they do not. So this goes a bit deeper than where the fees display. My two posts with the screenshots should cover it all.

**Comment by cayci** [comment]
Awesome!  I shall be sifting through the code abyss.  I'll be back  :)

**Comment by cayci** [comment]
I love the nice, clean codebase!  I'm a bit skittish until I know my way around.  Don't write me off... I'll be back after some sleep  :)

**Comment by cayci** [comment]
Hey, guys.  I'm struggling a tad.  Please help?

**Comment by cayci** [comment]
On a Mac.  Using MAMP as my local server.  Setting up the wp-config.php file.  I have a brainfart as to what I plug in for Database name, Database username, Database password, & Database host to view changes on my local machine.  I know the answer is probably right in front of me, but I'm spinning my wheels in the mud  :/       May need insight via a stroll on the Harley.  I'll be back...

**Comment by cayci** [comment]
Facepalms are welcome if you can point me in the right direction  ;)

**Comment by michaelbeil** [comment]
```php
/** MySQL database username */
define('DB_USER', 'root');

/** MySQL database password */
define('DB_PASSWORD', 'root');

/** MySQL hostname */
define('DB_HOST', 'localhost');
```

**Comment by cayci** [comment]
Michael... you are my hero.  I figured I was over-thinking this.  Facepalms away  :)

**Comment by cayci** [comment]
Totally on my end, with nothing to do with the repo I pull in, correct?

**Comment by michaelbeil** [comment]
I am not sure what the question is.

**Comment by cayci** [comment]
It's all good.  I've got it.  I get lost in small details sometimes....

Thank you for being nice about it  :)

**Comment by cayci** [comment]
Please tell me if I'm thinking about this correctly.  (I won't be this high-maintenance once I gain confidence.  I find open-source contributions to be intimidating, as I have in the past been called names for asking dumb questions.)   

The changes I will be making are:
Calculations affecting whether or not a fee is included in the subtotal will be made within easy-digital-downloads/includes/cart/functions.php.

Changes to the display, i.e, whether a fee is displayed within the top section of the cart (item fees), or below the subtotal in the footer rows (fees), will be made within easy-digital-downloads/includes/cart/template.php.

The subtotal should only include the total of each item-type fee times the quantity of that item.

Beneath the subtotal, there should be the following rows:
Total Discounts
Total Fees (Itemized, i.e., handling fee, shipping fee, etc?  Fees should not be taxed, correct?)
Total Taxes

Total

Thank you for bearing with me.  I promise to chill out once I get comfortable...   :)

**Comment by mintplugins** [comment]
@cayci It looks like you are on the right track but would need to do extensive testing to make sure it covers all possibilities. The process from here is basically trial and error with making code changes and testing them. If it works for you in your tests, send the updated code to the issue here and we can run tests with it as well. If you include screenshots of it working that's even better! (Ideally something that matches up to @SDavisMedia's screenshots above) 

If it doesn't work in your tests quite right yet, take another shot at changing the code again until your tests work the way you want them to.

To keep the issue clean, I’m gonna go ahead and delete all of the above replies that are directly related to the development and functionality. 

Thanks for getting involved!

**Comment by cayci** [comment]
Thank you so much   @mintplugins   :)

**Comment by cayci** [comment]
@SDavisMedia  Hi, Sean.  I saw your issue with the EDD Wallet Deposit, and how it doesn't reflect in the cart subtotal.  Would you please point me to the codebase for this, so I can see how everything integrates together?  It's not part of the main EDD repo, correct?   I'll work on the simpler stuff first, but don't want to ignore your issue...

**Comment by cayci** [comment]
Minor issue, and not relevant to this enhancement, but I found this broken link while digging around.  If you are on the WP EDD page:
https://wordpress.org/plugins/easy-digital-downloads/
... and you scroll down to "Developer friendly with dozens of actions and filters. View Codex"
The 'view codex' link is broken (https://easydigitaldownloads.com/codex/index.html)...
Sorry to be ornery  :)

**Comment by SDavisMedia** [comment]
Hi @cayci! That is correct you won't find any item type fees in any standard EDD core functionality. I'm gonna shoot you and email to the email address on your GitHub profile with a copy of EDD Wallet. Then we can communicate from there about anything that isn't directly related to the development of this issue, just to keep it tidy.

And thanks for catching the broken link! Will get that fixed.

**Comment by michaelbeil** [comment]
Request: https://secure.helpscout.net/conversation/274131302/46223/?folderId=180499

**Comment by cklosowski** [comment]
@SDavisMedia do we need to do much more with this? I'd love to have this one of the earliest major updates to `2.7` so we can make sure that it get's tested throughly.

**Comment by SDavisMedia** [comment]
@cklosowski Nope! I'll start playing with this and see what I can come up with. I'll ping you folks if I need help.

**Comment by SDavisMedia** [comment]
It looks like handling this issue is going to expose a lot of other flaws. I'm going to try to make sure it all gets taken care of, whether I do it here in this issue or if I have to request changes in other issues. I'm gonna be pushing code up but don't consider it done until it's all the way _done_. I keep running into new walls that make me go back and rework things. Never hesitate to chime in. 👌

**Comment by SDavisMedia** [comment]
Current state of affairs in a very basic cart configuration with taxes, prices exclusive of taxes, tax info displayed at checkout, a `fee` type fee, and a discount:

<img width="726" alt="screenshot 2016-11-20 09 58 25" src="https://cloud.githubusercontent.com/assets/2359131/20464982/20dcd0ce-af08-11e6-9c8a-ed8f0475bf7d.png">

This is only the beginning, though. This runs _much_ deeper than checkout cart display.

**Comment by chriscct7** [comment]
Just a sidebar I think we should denote top or bottom by a `fee checkout location`. That leaves us room to introduce actual fee types such as the needed condensed fee (open ticket about this) which  is an actual type of fee.

**Comment by pippinsplugins** [comment]
@SDavisMedia in what cases does `adjusted by discount(s)` show? Anytime there is a negative fee or discount code?

**Comment by SDavisMedia** [comment]
@pippinsplugins only when an EDD core discount is applied to the cart. It will "come and go" as a discount is added or removed. It is also toggled by the "Display during checkout" setting alongside the tax info displayed next to the cart item price.

**Comment by SDavisMedia** [comment]
@chriscct7 Is there a need to denote fee location if the fee types are overall treated differently? As in `fee` type fees are cart adjustments while `item` type fees are products.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by SDavisMedia** [comment]
@pippinsplugins sorry, I misspoke. It's true that the `adjusted by discount(s)` text only displays when a core discount is added. But it is _not_ controlled by that setting. I was thinking of the `plus n% tax` displayed below the fee, as seen here: https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4833#issuecomment-261793794

**Comment by chriscct7** [comment]
I think so, because there's different types of fees that will be in
different locations.

For example, a combined vs noncombined fee could be in either location.

**Comment by sunnyratilal** [comment]
@SDavisMedia Is this close to completion?

**Comment by SDavisMedia** [comment]
@sunnyratilal No, this needs to be punted. But I'll hit you in Slack about the reason why.

**Comment by sunnyratilal** [comment]
As per discussion with @SDavisMedia and @cklosowski, this is being punted to 2.8. This issue brings out other discrepancies across EDD which would all need to addressed at once.

**Comment by SDavisMedia** [comment]
@sunnyratilal I don't think it's a good idea to have this in 3.0, unless my perception of the 3.0 timeline is off. I don't want to rush this at all and I want as much testing and feedback as possible. I'd have to start fresh now that the data source is different, though I'd probably go down the same design path. But that was almost 2 years ago... so I want to approach it fresh again.

If this can wait until shortly after 3.0, I think that would be best.

**Comment by dgoldak** [comment]
Customer request, reporting the subtotal is confusing when a fee (recurring fee) is in the cart.
https://secure.helpscout.net/conversation/1233005582/160695
